File: 2023/Day 9/php/day9.php
<?php

function differences($seq) {
    $diffs = [];
    for ($i = 0; $i < count($seq) - 1; $i++) {
        $diffs[] = $seq[$i + 1] - $seq[$i];
    }
    return $diffs;
}

function buildSequences($seq) {
    $sequences = [$seq];
    while (!allZeros(end($sequences))) {
        $sequences[] = differences(end($sequences));
    }
    return $sequences;
}

function allZeros($seq) {
    foreach ($seq as $num) {
        if ($num !== 0) return false;
    }
    return true;
}

function solveSequence($seq) {
    $sequences = buildSequences($seq);
    return array_sum(array_map(fn($s) => end($s), $sequences));
}

function solveFromFile($filename) {
    $lines = file($filename, FILE_IGNORE_NEW_LINES);
    $partOneSum = array_sum(array_map(fn($line) => solveSequence(array_map('intval', explode(' ', $line))), $lines));
    $partTwoSum = array_sum(array_map(fn($line) => solveSequence(array_reverse(array_map('intval', explode(' ', $line)))), $lines));
    return [$partOneSum, $partTwoSum];
}

[$p1, $p2] = solveFromFile("input.txt");
echo "Part One Result: $p1<br>Part Two Result: $p2<br>";



// Smaller version
function solveSequenceA($seq) {
    $sequences = [$seq];

    while (end($sequences)) {
        $lastSeq = end($sequences);
        if (count(array_unique($lastSeq)) === 1 && end($lastSeq) === 0) break;
        $newSeq = [];
        for ($i = 0; $i < count($lastSeq) - 1; $i++) {
            $newSeq[] = $lastSeq[$i + 1] - $lastSeq[$i];
        }
        $sequences[] = $newSeq;
    }

    return array_sum(array_map(fn($s) => end($s), $sequences));
}

function solveFromFileA($filename) {
    $lines = file($filename, FILE_IGNORE_NEW_LINES);
    $partOneSum = array_sum(array_map(fn($line) => solveSequenceA(array_map('intval', explode(' ', $line))), $lines));
    $partTwoSum = array_sum(array_map(fn($line) => solveSequenceA(array_reverse(array_map('intval', explode(' ', $line)))), $lines));
    return [$partOneSum, $partTwoSum];
}

[$p1, $p2] = solveFromFileA("input.txt");
echo "Part One Result: $p1<br>Part Two Result: $p2<br>";
?>

File: 2023/Day 9/perl/day9.pl
#!/usr/bin/perl
use strict;
use warnings;
use List::Util qw(sum);

sub differences {
    my @seq = @_;
    return map { $seq[$_ + 1] - $seq[$_] } 0 .. $#seq - 1;
}

sub build_sequences {
    my @seq = @_;
    if (@seq == grep { $_ == 0 } @seq) {
        return ([@seq]);
    } else {
        return ([@seq], build_sequences(differences(@seq)));
    }
}

sub solve_sequence {
    my @seq = @_;
    my @sequences = build_sequences(@seq);
    return sum(map { $_->[-1] } @sequences);
}

sub solve_from_file {
    my ($filename) = @_;
    my $p1 = 0;
    my $p2 = 0;

    open my $fh, '<', $filename or die "Cannot open file: $!";
    while (my $line = <$fh>) {
        chomp $line;
        my @numbers = split / /, $line;
        $p1 += solve_sequence(@numbers);
        $p2 += solve_sequence(reverse @numbers);
    }
    close $fh;

    return ($p1, $p2);
}

my ($p1, $p2) = solve_from_file("input.txt");
print "Part One Result: $p1\n";
print "Part Two Result: $p2\n";



# Streamlined Version

sub solve_sequence_a {
    my @seq = @_;
    my @sequences = ([@seq]);

    while (my @last_seq = @{$sequences[-1]}) {
        last if @last_seq == grep { $_ == 0 } @last_seq;
        push @sequences, [map { $last_seq[$_ + 1] - $last_seq[$_] } 0 .. $#last_seq - 1];
    }

    return sum(map { $_->[-1] } @sequences);
}

sub solve_from_file_a {
    my ($filename) = @_;
    open my $fh, '<', $filename or die "Cannot open file: $!";
    my ($p1, $p2) = (0, 0);

    while (my $line = <$fh>) {
        chomp $line;
        my @numbers = split / /, $line;
        $p1 += solve_sequence_a(@numbers);
        $p2 += solve_sequence_a(reverse @numbers);
    }
    close $fh;

    return ($p1, $p2);
}

# Example usage
my ($p1a, $p2a) = solve_from_file_a("input.txt");
print "Part One Result: $p1a\n";
print "Part Two Result: $p2a\n";
